### PR TITLE
DOC: add docs for the main namespace

### PR DESCRIPTION
**Pull request by rgommers**
This PR was triggered by `show_config` missing from the docs completely (xref gh-17862). The main namespace does not contain much functionality, but it should have a page which lists what is public API in the main namespace.

The regular `.. autosummary` thing unfortunately wasn't usable here, Sphinx makes a mess with duplicate labels and docstrings. So I used a plain reST table for most entries.

The PR grew a bit - also noticed that `PytestTester` needed a docstring. And I'm commenting out some noisy output of `dev.py` that didn't land in the right place.

I'll note that the doc build will likely fail until gh-17862 is merged, because the current `show_config` docstring contains one invalid reference. I didn't fix that to avoid merge conflicts with gh-17862.

### CODE REVIEW

File: doc/source/reference/main_namespace.rst
Comment (tupui) [inline doc/source/reference/main_namespace.rst:40]: I am not sure about keeping legacy and deprecated things. Legacy could make sense, but deprecated a bit less to me.

File: scipy/_lib/_testutils.py
Comment (tupui) [inline scipy/_lib/_testutils.py:29]: This section might be misleading as it's not the parameters for the init but the call method. Maybe adding a small example or a one liner explanation would help?

**Review by tupui — APPROVED** [review]
Thanks Ralf, I have a few suggestions but otherwise LGTM (provided we get the other PR in as you pointed out.)

File: scipy/_lib/_testutils.py
Comment (rgommers) [reply scipy/_lib/_testutils.py:29]: I considered that and actually started writing a simple example using `scipy.test` - before realizing that that would be extraordinarily expensive if the doctests are run:)

Comment (tupui) [reply scipy/_lib/_testutils.py:29]: Right 👍

File: doc/source/reference/main_namespace.rst
Comment (rgommers) [reply doc/source/reference/main_namespace.rst:40]: I'd like to reject this comment @tupui, and keep things as they are. We have it in the API reference and in the sidebar. Once the deprecation expires it can be cleaned up. This is standard for all deprecated functionality - no reason to deviate here.

**Review by ev-br — COMMENTED** [review]
Two optional suggestions

**Pull request merged**
